Mathlib/Data/Nat/Factorization/Multiplicity.lean

```diff
@@ -0,0 +1,309 @@
+        simp only [mem_Ico, Nat.lt_succ_iff,
+          fin.pow_dvd_iff_le_multiplicity, Finset.mem_filter,
+          and_assoc, and_congr_right_iff, iff_and_self]
+        intro hi h
+        rw [← fin.pow_dvd_iff_le_multiplicity] at h
+        rcases m with - | m
+        · rw [zero_pow, zero_dvd_iff] at h
+          exacts [(hn.ne' h).elim, one_le_iff_ne_zero.1 hi]
+        refine LE.le.trans_lt ?_ hb
+        exact le_log_of_pow_le (one_lt_iff_ne_zero_and_ne_one.2 ⟨m.succ_ne_zero, hm⟩)
+          (le_of_dvd hn h)
+
+theorem factorization_pow_self {p n : ℕ} (hp : p.Prime) : (p ^ n).factorization p = n := by
+  simp only [factorization_pow, coe_smul, Pi.smul_apply, smul_eq_mul]
+  have h1: p.factorization p = 1 := by
+    exact Prime.factorization_self hp
+  rw [h1]
+  simp
+
+/-- **Legendre's Theorem**
+
+The multiplicity of a prime in `n!` is the sum of the quotients `n / p ^ i`. This sum is expressed
+over the finset `Ico 1 b` where `b` is any bound greater than `log p n`. -/
+theorem factorization_factorial {p : ℕ} (hp : p.Prime) :
+  ∀ {n b : ℕ}, log p n < b → (n !).factorization p = (∑ i ∈ Ico 1 b, n / p ^ i : ℕ)
+  | 0, b, _ => by simp only [factorial_zero, factorization_one, coe_zero, Pi.zero_apply,
+    Nat.zero_div, sum_const_zero]
+  | n + 1, b, hb =>
+    calc
+      ((n + 1)!).factorization p = ( (n + 1) * (n !)).factorization p := by
+        rw [← factorial_succ]
+      _ = (n + 1).factorization p + (n !).factorization p := by
+        rw [factorization_mul (zero_ne_add_one n).symm (factorial_ne_zero n), coe_add, Pi.add_apply]
+      _ = #{i ∈ Ico 1 b | p ^ i ∣ n + 1} + (∑ i ∈ Ico 1 b, n / p ^ i : ℕ) := by
+        rw [factorization_factorial hp ((log_mono_right <| le_succ _).trans_lt hb)]
+        simp only [add_left_inj]
+        apply factorization_eq_card_pow_dvd₀ (Prime.ne_one hp) hp (zero_lt_succ n)
+        exact hb
+      _ = (∑ i ∈ Ico 1 b, n / p ^ i : ℕ) + #{i ∈ Ico 1 b | p ^ i ∣ n + 1} := by
+        exact
+          Nat.add_comm (#(Finset.filter (fun i ↦ p ^ i ∣ n + 1) (Ico 1 b)))
+            (∑ i ∈ Ico 1 b, n / p ^ i)
+      _ = (∑ i ∈ Ico 1 b, (n / p ^ i + if p ^ i ∣ n + 1 then 1 else 0) : ℕ) := by
+        rw [sum_add_distrib, sum_boole]
+        simp
+      _ = (∑ i ∈ Ico 1 b, (n + 1) / p ^ i : ℕ) :=
+        Finset.sum_congr rfl fun _ _ => (succ_div _ _).symm
+
+/-- For a prime number `p`, taking `(p - 1)` times the factorization of `p` in `n!` equals `n` minus
+the sum of base `p` digits of `n`. -/
+theorem sub_one_mul_factorization_factorial {n p : ℕ} (hp : p.Prime) :
+    (p - 1) * (n !).factorization p = n - (p.digits n).sum := by
+  simp only [factorization_factorial hp <| lt_succ_of_lt <| lt.base (log p n),
+    ← Finset.sum_Ico_add' _ 0 _ 1, Ico_zero_eq_range, ←
+    sub_one_mul_sum_log_div_pow_eq_sub_sum_digits]
+
+/-- Modified version of `factorization_prod` that accounts for inputs. -/
+theorem factorization_prod₀ {α : Type*} {p : ℕ}
+{S : Finset α} {g : α → ℕ} (hS : ∀ x ∈ S, g x ≠ 0) :
```

Comment:
Here, it would be better to follow the [style guidelines](https://leanprover-community.github.io/contribute/style.html) for indentation:)